lib/rules/utils/ast-utils.js

```diff
@@ -602,24 +602,32 @@ module.exports = {
      *
      * First, this checks the node:
      *
-     * - The function name does not start with uppercase (it's a constructor).
+     * - The function name does not start with uppercase. It's a convention to capitalize the names
+     *   of constructor functions. This check is not performed if `capIsConstructor` is set to `false`.
      * - The function does not have a JSDoc comment that has a @this tag.
      *
      * Next, this checks the location of the node.
      * If the location is below, this judges `this` is valid.
      *
      * - The location is not on an object literal.
-     * - The location is not assigned to a variable which starts with an uppercase letter.
+     * - The location is not assigned to a variable which starts with an uppercase letter. Applies to anonymous
+     *   functions only, as the name of the variable is considered to be the name of the function in this case.
+     *   This check is not performed if `capIsConstructor` is set to `false`.
      * - The location is not on an ES2015 class.
      * - Its `bind`/`call`/`apply` method is not called directly.
      * - The function is not a callback of array methods (such as `.forEach()`) if `thisArg` is given.
      *
      * @param {ASTNode} node - A function node to check.
      * @param {SourceCode} sourceCode - A SourceCode instance to get comments.
+     * @param {boolean} [capIsConstructor = true] - `false` disables the assumption that functions which name starts
+     * with an uppercase or are assigned to a variable which name starts with an uppercase are constructors.
      * @returns {boolean} The function node is the default `this` binding.
      */
-    isDefaultThisBinding(node, sourceCode) {
-        if (isES5Constructor(node) || hasJSDocThisTag(node, sourceCode)) {
+    isDefaultThisBinding(node, sourceCode, { capIsConstructor = true } = {}) {
```

Comment:
I wonder if instead of having a `capIsContructor` option (which feels like implementation details leaking), we could have the option be called something like `allowES5Constructor` (defaulting to `true`)? It would require some refactoring within this method, but I think that's a better API for this util. Thoughts?

Reply:
I'm not sure, because of this - a function can be an ES5 constructor regardless of whether its name starts with an uppercase, i.e. any function meant to be called with `new` is an ES5 constructor?

For instance, `isDefaultThisBinding()` or even better `isES5Constructor()` could in the future also check for the [class tag](https://jsdoc.app/tags-class.html) (`@class` or `@constructor` synonyms) to find out that `this` is allowed (i.e. not default binding) even if the function starts with a lowercase. 

If the param has a generic `'...ES5Constructor'` name , `false` would mean disabling all checks that could find out that the function is an ES5Constructor, which probably isn't useful as disabling just the uppercase convention. The check for an uppercase alone looks to me more like a configurable specification detail than an implementation detail of the function.

What certainly doesn't look good is the `(capIsConstructor && isES5Constructor(node))` condition few lines below. It would be better to pass `capIsConstructor` to `isES5Constructor()`. Though, it might look strange since that's the only check the `isES5Constructor()` is doing at the moment.

Reply:
The post was confusing, I'll try to clarify what I think,  

This rule and the helper function intend to find invalid use of `this`. In constructor functions, the use of `this` is valid. If there would be another way to find out that a function is certainly used as a constructor function ('ES5' constructor), I don't think there should be an option to disable that check. Such an option would indirectly disallow constructor functions, and the rule would be more like a  'prefer-class-syntax' rule.

That's why I think that the `capIsConstructor` option's name is correct for its purpose because the intention is not to disallow ES5 constructors, but just to 'disable' the assumption that a function that starts with an uppercase is an ES5 constructor.

Reply:
@kaicataldo maybe we could continue with this discussion :-)

Reply:
Sorry for losing track of this! Thanks for the explanation. That makes sense to me!